## Упражнения по библиотеке Numpy

In [2]:
import numpy as np

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [70]:
arr = np.random.randint(0, 10, 10)
arr = np.where((arr > 3) & (arr < 8), -arr, arr)
arr

array([ 3, -5,  8, -7,  2,  1,  0,  3, -5,  1])

**2.** Заменить максимальный элемент случайного массива на 0

In [73]:
arr = np.random.randint(0, 100, 10)
arr[np.where(arr == np.max(arr))] = 0
arr

array([78,  3,  0,  7, 60,  5, 11, 71,  8, 82])

**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [231]:
arrays = [np.asarray(a) for a in ([1, 2, 3], [4, 5], [6, 7])]
indexes = np.indices(map(len, arrays), dtype=int).reshape(len(arrays), -1).T

for n, arr in enumerate(arrays):
    indexes[:, n] = arrays[n][indexes[:, n]]

indexes

array([[1, 4, 6],
       [1, 4, 7],
       [1, 5, 6],
       [1, 5, 7],
       [2, 4, 6],
       [2, 4, 7],
       [2, 5, 6],
       [2, 5, 7],
       [3, 4, 6],
       [3, 4, 7],
       [3, 5, 6],
       [3, 5, 7]])

**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [241]:
A = np.random.randint(0, 5, (8, 3))
B = np.random.randint(0, 5, (2, 2))

print(A, B, sep='\n\n')

C = (A[..., np.newaxis, np.newaxis] == B)
rows = (C.sum(axis = (1, 2, 3)) >= B.shape[1]).nonzero()[0]
print(rows)

[[1 3 0]
 [1 4 2]
 [4 2 1]
 [0 0 1]
 [3 1 3]
 [1 4 0]
 [2 1 3]
 [4 4 2]]

[[3 2]
 [0 3]]
[0 3 4 6]


**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [151]:
A = np.random.randint(0, 5, 30).reshape(10, 3)
mask = np.logical_and.reduce(A[:,1:] == A[:,:-1], axis=1) & np.logical_and.reduce(A[:,1:] == A[:,[0, 2]], axis=1)\
    & np.logical_and.reduce(A[:,:-1] == A[:,[0, 2]], axis=1)
A[~mask]

array([[4, 0, 0],
       [0, 0, 2],
       [0, 2, 0],
       [1, 0, 0],
       [4, 3, 2],
       [1, 3, 3],
       [0, 0, 4],
       [3, 4, 2],
       [0, 4, 2],
       [4, 3, 0]])

**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [50]:
A = np.random.randint(0, 5, 9).reshape(3, 3)
B = np.array([[1, 1, 1], [2, 2, 2], [1, 1, 1]])
unique1 = np.unique(A, axis=0)
unique2 = np.unique(B, axis=0)
unique1, A, unique2, B

(array([[2, 0, 4],
        [2, 4, 1],
        [3, 4, 3]]),
 array([[2, 0, 4],
        [3, 4, 3],
        [2, 4, 1]]),
 array([[1, 1, 1],
        [2, 2, 2]]),
 array([[1, 1, 1],
        [2, 2, 2],
        [1, 1, 1]]))

______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [121]:
X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])

if X.__len__() < X[0].__len__():
    n = X.__len__()
else:
    n = X[0].__len__()

prod = 1
for i in range(n):
    if X[i][i] != 0:
        prod *= X[i][i]

np.prod(np.diagonal(X)[~np.in1d(np.diagonal(X), 0)].reshape(1, 2)), prod

(3, 3)

* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [157]:
# your code

* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [82]:
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])
python_method = max([x[i] for i in range(x.__len__()) if i != 0 if x[i - 1] == 0])
numpy_method = np.max(x[1:][(x == 0)[:-1]])
python_method, numpy_method

(5, 5)

* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).


In [89]:
x = np.array([2, 2, 2, 3, 3, 3, 5])

list_of_num = []
list_of_repeat = []
for i in range(x.__len__()):
    k = 0
    
    if x[i] not in list_of_num:
        list_of_num.append(x[i])
    else:
        continue
    
    if i == x.__len__() - 1:
        list_of_repeat.append(k)
        break
    
    for j in range(i + 1, x.__len__()):
        if x[i] == x[j]:
            k += 1
        else:
            list_of_repeat.append(k)
            break
            
print(list_of_num, list_of_repeat)

[2, 3, 5] [2, 2, 0]


In [116]:
starts = np.r_[0, np.flatnonzero(~np.isclose(x[1:], x[:-1], equal_nan=True)) + 1]
lengths = np.diff(np.r_[starts, x.__len__()]) - 1
values = x[starts]
    
lengths, values

(array([2, 2, 0], dtype=int64), array([2, 3, 5]))

* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [174]:
def diff(a, b):
    result = []
    for i in range(a.__len__()):
        result.append(a[i] - b[i])
    return result

In [ ]:
%%time
from math import sqrt

a = np.random.randint(10, size=(25000, 4))
b = np.random.randint(10, size=(24000, 4))

result_list = []
for i in a:
    row = []
    for j in b:
        row.append(np.sqrt(sum(pow(a - b, 2) for a, b in zip(i, j))))
    result_list.append(row)

result_list

In [187]:
np.sqrt(np.array(sum([pow(np.array(diff(x, y)), 2) for x, y in zip(a, b)])))

array([9.        , 4.69041576, 5.56776436, 4.24264069])

In [203]:
%%time

a = np.random.randint(10, size=(22500, 4))
b = np.random.randint(10, size=(22400, 4))

P = np.add.outer(np.sum(a ** 2, axis=1), np.sum(b ** 2, axis=1))
N = np.dot(a, b.T)
print(np.sqrt(P - 2*N))

[[ 6.55743852  4.24264069  8.1240384  ...  9.2736185   8.1240384
  10.04987562]
 [ 9.2736185   4.79583152  9.32737905 ...  9.8488578   8.77496439
  12.24744871]
 [ 7.21110255  6.244998   12.28820573 ...  9.32737905  3.31662479
   6.32455532]
 ...
 [ 6.78232998  3.          8.06225775 ...  6.244998    5.19615242
   8.36660027]
 [ 4.58257569  5.47722558 10.67707825 ...  8.71779789  4.69041576
   4.79583152]
 [ 7.28010989  8.94427191  2.82842712 ...  8.1240384  12.24744871
  10.72380529]]
CPU times: total: 8.69 s
Wall time: 8.71 s


In [206]:
%%time
import scipy

a = np.random.randint(10, size=(22500, 4))
b = np.random.randint(10, size=(22400, 4))

scipy.spatial.distance.cdist(a, b)

CPU times: total: 4.64 s
Wall time: 4.64 s


array([[ 5.38516481,  3.60555128, 12.28820573, ..., 11.83215957,
         7.74596669,  9.89949494],
       [ 5.19615242,  5.56776436, 10.81665383, ..., 13.19090596,
         5.09901951,  6.        ],
       [ 7.68114575, 10.04987562,  7.54983444, ..., 11.04536102,
         6.4807407 ,  2.44948974],
       ...,
       [ 7.07106781, 10.        , 12.40967365, ..., 15.06651917,
         5.74456265,  8.06225775],
       [ 7.93725393,  8.06225775,  8.30662386, ..., 10.09950494,
         8.1240384 ,  5.47722558],
       [ 7.34846923, 10.48808848, 12.08304597, ..., 15.06651917,
         6.08276253,  7.41619849]])

_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [350]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [351]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [354]:
average_calories = calorie_stats.mean()
average_calories, average_calories - 60

(106.88311688311688, 46.883116883116884)

3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [357]:
calorie_stats_sorted = np.sort(calorie_stats)
calorie_stats_sorted

array([ 50.,  50.,  50.,  70.,  70.,  80.,  90.,  90.,  90.,  90.,  90.,
        90.,  90., 100., 100., 100., 100., 100., 100., 100., 100., 100.,
       100., 100., 100., 100., 100., 100., 100., 100., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 120., 120., 120., 120., 120., 120., 120.,
       120., 120., 120., 130., 130., 140., 140., 140., 150., 150., 160.])

4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [359]:
median_calories = np.median(calorie_stats_sorted)
median_calories

110.0

5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [368]:
np.percentile(calorie_stats_sorted, 90),\
np.percentile(calorie_stats_sorted, 60),\
np.percentile(calorie_stats_sorted, 40),\
np.percentile(calorie_stats_sorted, 20),\
np.percentile(calorie_stats_sorted, 10),\
np.percentile(calorie_stats_sorted, 3.5)

(124.00000000000006, 110.0, 110.0, 100.0, 90.0, 63.2)

6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [372]:
more_calories = calorie_stats_sorted[calorie_stats_sorted > 60].__len__() / calorie_stats_sorted.__len__()
more_calories

0.961038961038961

7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

Данные которые мы рассматриваем могут быть очень разбросаны и теряется вес среднего значения. Как видим разброс всего 19.36, так что можем утверждать что данные кучно расспределенны.

In [374]:
calorie_std = np.std(calorie_stats_sorted)
calorie_std

19.35718533390827

8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

На коробке написать надпись "Наши хлопью менее калорийные чем у 96% конкурентов" или что-то похожее 